### PR TITLE
Implemented Staticline GUI auto-layout from config files  

### PR DESCRIPTION
**Pull request by huanyanqi**
Created functionality to automatically generate and lay out staticline GUIs based on a config file that specifies a list of devices and their control type. The functionality for the GUI buttons are also automatically assigned and handled by the hardware handler. Config files are also integrated in the setup of the hardware handler functions to specify parameters (e.g. `up_voltage`) that are used by the Staticline.

### Demo of operation
https://streamable.com/q1jzk8

### Sample staticline GUI config file
```
{ 
    "1": {
        "type": "digital",
        "name": "Digital staticline device 1",
        "hardware_type" : "abstract"
    },
    "2": {
        "type": "analog",
        "name": "Analog staticline device 1",
        "hardware_type" : "abstract2"
    }
}
```
The `type` field can only be `"digital"` or `"analog"`, and the `hardware_type` is the name of the hardware's device server.

### Adding new devices
New devices can be added by adding a new hardware handler class for that device in `staticline_devices.py` which implements the functions `self.up/self.down` for digital lines or `self.set_value` for analog lines. An entry for that device then needs to be added in the `registered_staticline_modules` dictionary (key = name of the device server, value = hardware handler class). 

### Known Issues / Future Improvements
- Currently, each device is specified using its hardware type, which is the name of its device server. This prevents us from having >1 device per device hardware model. 
- Allow the config file to specify >1 static line per device

### CODE REVIEW

File: pylabnet/hardware/staticline/staticline_devices.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
+from abc import ABC, abstractmethod
```
Comment (cknaut) [inline pylabnet/hardware/staticline/staticline_devices.py:1]: Cool use of `abc`s, definitely a more elegant implementation than before.

File: pylabnet/launchers/servers/staticline_nidaqmx.py [hunk 2]
```diff
@@ -2,10 +2,11 @@
 
 import socket
 
-import pylabnet.hardware.ni_daqs.nidaqmx_card as nidaqmx
+# import pylabnet.hardware.ni_daqs.nidaqmx_card as nidaqmx
 import pylabnet.hardware.staticline.staticline as staticline
 
 from pylabnet.network.client_server.staticline import Service, Client
+import pylabnet.network.client_server.abstract_device as abstract_device
```
Comment (cknaut) [inline pylabnet/launchers/servers/staticline_nidaqmx.py:9]: I don't think you have actually included the `abstract_device` servers in the directory, which means this dummy example can not be run. I suggest either adding `abstract_device` to the servers directory, or using the `dummy=True` flag in order to instantiate devices as indicated here: https://github.com/lukingroup/pylabnet/pull/122

**Review by cknaut — CHANGES_REQUESTED** [review]
This branch includes an overall very nice implementation of a generic Staticline GUI launcher. The core functionality of programmatically building the GUI as well as the rewritten `StaticLine` module looks great. I was able to instantiate a DIO toggle for our Zurich Instrument HDAWG by using this config-file:

```json
{
  "1": {
    "type": "digital",
    "name": "HDAWG DIO bit 16",
    "hardware_type": "zi_hdawg",
    "DIO_bit": 16
  }
}
```

The resulting GUI worked properly:
Bit high:
image

Bit low:
image


While trying to test the code using NIDAQs and multiple devices, I quickly ran into the limitations that you've already mentioned in the pull request:

- Addressing multiple devices of the same type is not straightforward.
- Accessing the same device for two different `StaticLines` is not supported.

In order for this branch to be fully integrated into the experimental workflow, these two topics will have to be addressed. At this stage, this PR is ready to merge, once the `abstract_device` has been added (see my other comment).

File: pylabnet/launchers/servers/staticline_nidaqmx.py (hunk 2 above)
Comment (huanyanqi) [reply pylabnet/launchers/servers/staticline_nidaqmx.py:9]: This file is no longer in use as we no longer require the implementation of a device-specific Statcline server, and so this file can be deleted. Instead, we only need the normal device server to be running in order for the Staticline GUI to communicate with the device.,

**Pull request merged**